app/src/main/java/org/glucosio/android/presenter/AddGlucosePresenter.java

```diff
@@ -70,49 +71,40 @@ public int hourToSpinnerType(int hour) {
     }
 
     public void dialogOnAddButtonPressed(String time, String date, String reading, String type, String notes) {
-        if (validateDate(date) && validateTime(time) && validateGlucose(reading) && validateType(type)) {
-            Date finalDateTime = getReadingTime();
-            boolean isReadingAdded;
-            if ("mg/dL".equals(getUnitMeasuerement())) {
-                int finalReading = Integer.parseInt(reading);
-                GlucoseReading gReading = new GlucoseReading(finalReading, type, finalDateTime, notes);
-                isReadingAdded = dB.addGlucoseReading(gReading);
-            } else {
-                converter = new GlucosioConverter();
-                int convertedReading = converter.glucoseToMgDl(Double.parseDouble(reading));
-                GlucoseReading gReading = new GlucoseReading(convertedReading, type, finalDateTime, notes);
-
-                isReadingAdded = dB.addGlucoseReading(gReading);
-            }
-            if (!isReadingAdded) {
-                activity.showDuplicateErrorMessage();
-            } else {
-                activity.finishActivity();
-            }
-        } else {
-            activity.showErrorMessage();
-        }
+        dialogOnAddButtonPressed(time, date, reading, type, notes, UNKNOWN_ID);
     }
 
     public void dialogOnAddButtonPressed(String time, String date, String reading, String type, String notes, long oldId) {
         if (validateDate(date) && validateTime(time) && validateGlucose(reading) && validateType(type)) {
             Date finalDateTime = getReadingTime();
-            boolean isReadingAdded;
-            if ("mg/dL".equals(getUnitMeasuerement())) {
-                int finalReading = Integer.parseInt(reading);
-                GlucoseReading gReading = new GlucoseReading(finalReading, type, finalDateTime, notes);
-                isReadingAdded = dB.editGlucoseReading(oldId, gReading);
-            } else {
-                converter = new GlucosioConverter();
-                int convertedReading = converter.glucoseToMgDl(Double.parseDouble(reading));
-                GlucoseReading gReading = new GlucoseReading(convertedReading, type, finalDateTime, notes);
-
-                isReadingAdded = dB.editGlucoseReading(oldId, gReading);
-            }
-            if (!isReadingAdded) {
-                activity.showDuplicateErrorMessage();
+            Number number = ReadingTools.parseReading(reading);
+            if (number == null) {
+                activity.showErrorMessage();
             } else {
-                activity.finishActivity();
+                boolean isReadingAdded;
+                if ("mg/dL".equals(getUnitMeasuerement())) {
```

Comment:
That is the third nested level in the logic flow. How to about extract some logic to methods (like `createReading`) to make it more readable?